### PR TITLE
Adding quantized::conv2d function for pytorch mobile in c10

### PR DESCRIPTION
**Pull request by supriyar**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #26335 Unify Quantization APIs for add, pool and relu
* #26307 Changes to support int8 weight and fp32 bias in QNNPACK
* #26211 Add support to call unpack for pytorch mobile quantized FC and Conv
* **#26152 Adding quantized::conv2d function for pytorch mobile in c10**

Summary:
This change adds the support to call QNNPACK using the refactored API for Conv2d operators

Test Plan:
python test/test_quantized.py TestQNNPackOps.test_qconv_qnnpack

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D17459892](https://our.internmc.facebook.com/intern/diff/D17459892)

### CODE REVIEW

File: test/test_quantized.py [hunk 1]
```diff
@@ -1322,6 +1323,129 @@ def test_qlinear_qnnpack(self, batch_size, input_channels, output_channels, use_
...
+
+        result_NHWK = result_ref.permute([0, 2, 3, 1])
+        if use_relu:
+            relu = torch.nn.ReLU()
+            result_NHWK = relu(result_NHWK)
+
+        result_ref_q = torch.quantize_linear(result_NHWK, scale=Y_scale, zero_point=Y_zp, dtype=torch.quint8)
+
+        np.testing.assert_array_almost_equal(result_ref_q.int_repr().numpy(), Y_q.int_repr().numpy(), decimal=0)
+
```
Comment (raghuramank100) [inline test/test_quantized.py]: Please also add a test with qnnpack to : https://github.com/pytorch/pytorch/blob/master/test/test_quantized_models.py

Comment (supriyar) [reply test/test_quantized.py]: I will add in a separate PR as we also need to add support to accept int8 weights in QNNPACK.

File: aten/src/ATen/native/quantized/cpu/qconv_prepack.cpp [hunk 2]
```diff
@@ -127,7 +133,112 @@ class QConvPackWeightInt8 final : public c10::OperatorKernel {
...
+  Tensor operator()(
+      Tensor weight,
+      c10::optional<Tensor> bias,
+      torch::List<int64_t> stride,
+      torch::List<int64_t> padding,
+      torch::List<int64_t> dilation,
+      int64_t groups) {
+    auto& ctx = at::globalContext();
+#ifdef USE_FBGEMM
+    if (ctx.preferredQuantizedEngine() == at::QEngine::FBGEMM) {
```
Comment (dzhulgakov) [inline aten/src/ATen/native/quantized/cpu/qconv_prepack.cpp]: what is the default for the preferred engine? is it going to be set correctly in the mobile build (based on the compile flags)

Comment (supriyar) [reply aten/src/ATen/native/quantized/cpu/qconv_prepack.cpp]: Default is currently set to FBGEMM. I can set the default for mobile build in a separate PR. Right now I'm setting it using `torch.backends.quantized.engine = torch.qnnpack` in the tests.

File: aten/src/ATen/native/quantized/cpu/qconv.cpp [hunk 3]
```diff
@@ -111,12 +113,31 @@ class QConv2dInt8 final : public c10::OperatorKernel {
     int kernel_h = kernel[0];
     int kernel_w = kernel[1];
 
-    TORCH_CHECK(C == (packB->inputChannels()),
-        "[QConv2D] Given groups=", groups, ", weight of size ",
-        K, ", ",  kernel_h, ", ", kernel_w, ", ", packB->inputChannels(),
-        ", expected input (NHWC) ", N, ", ", H, ", ", W, ", ", C,
-        " to have ", (packB->inputChannels() * groups),
-        " channels, but got ", C, " channels instead");
+    TORCH_CHECK(
```
Comment (dzhulgakov) [inline aten/src/ATen/native/quantized/cpu/qconv.cpp]: nit: maybe disable clang format for it - it looks unreadable:

```
// clang-format off
TORCH_CHECK(...
// clang-format on

File: aten/src/ATen/native/quantized/cpu/qconv.cpp [hunk 4]
```diff
@@ -244,6 +265,155 @@ class QConv2dInt8 final : public c10::OperatorKernel {
...
+  at::Tensor qnnpack_conv(
+      Tensor act,
+      Tensor packed_weight,
+      torch::List<int64_t> stride,
+      torch::List<int64_t> padding,
+      torch::List<int64_t> dilation,
+      int64_t groups,
+      double output_scale,
+      int64_t output_zero_point) {
+    TORCH_CHECK(
```
Comment (dzhulgakov) [inline aten/src/ATen/native/quantized/cpu/qconv.cpp]: extract these checks in to a subfunction and share with fbgemm implementation?

File: aten/src/ATen/native/quantized/cpu/qconv.cpp [hunk 5]
```diff
@@ -244,6 +265,155 @@ class QConv2dInt8 final : public c10::OperatorKernel {
...
+        act.ndimension() == 4,
+        "quanitzed::conv2d (qnnpack): Expected activation tensor to be 4-dimensional");
```
Comment (dzhulgakov) [inline aten/src/ATen/native/quantized/cpu/qconv.cpp]: quantized typo

File: aten/src/ATen/native/quantized/cpu/qconv.cpp [hunk 6]
```diff
@@ -244,6 +265,155 @@ class QConv2dInt8 final : public c10::OperatorKernel {
...
+            outShape.begin(), outShape.end(), [](int64_t i) { return i > 0; }),
+        "quanitzed::conv2d (qnnpack): each dimension of output tensor should be greater "
+        "than 0")
+    TORCH_CHECK(
+        (outShape[3] == out_ch),
+        "quanitzed::conv2d (qnnpack): Number of filters must be equal to number of "
+        "output channels")
+
+    // Allocate output Tensor and a buffer for QNNPACK to use
+    Tensor output = at::_empty_affine_quantized(
```
Comment (dzhulgakov) [inline aten/src/ATen/native/quantized/cpu/qconv.cpp]: we don't support per channel quant. Do you have check about it somewhere?

File: aten/src/ATen/native/quantized/cpu/qconv.cpp [hunk 7]
```diff
@@ -244,6 +265,155 @@ class QConv2dInt8 final : public c10::OperatorKernel {
...
+        nullptr);
+
+    TORCH_INTERNAL_ASSERT(
+        runStatus == pytorch_qnnp_status_success,
+        "failed to run quanitzed::conv2d (qnnpack) operator");
+
+    return output;
+  }
+#endif
+#if defined(USE_FBGEMM) || defined(USE_PYTORCH_QNNPACK)
```
Comment (dzhulgakov) [inline aten/src/ATen/native/quantized/cpu/qconv.cpp]: I think you don't need this #ifdef. - you can just have one function and it'd throw if neither of them is available

If it's about warnings that arguments are not used - I think there's a way to disable the warnings.

File: aten/src/ATen/native/quantized/cpu/qconv.cpp [hunk 8]
```diff
@@ -244,6 +265,155 @@ class QConv2dInt8 final : public c10::OperatorKernel {
...
+          padding,
+          dilation,
+          groups,
+          output_scale,
+          output_zero_point);
+    }
+#endif
+    TORCH_INTERNAL_ASSERT(
+        "Didn't find engine for operation quantized::conv ",
+        toString(ctx.preferredQuantizedEngine()));
```
Comment (dzhulgakov) [inline aten/src/ATen/native/quantized/cpu/qconv.cpp]: would be nice to have a constant declared somewhere for which engines are built in and print it here

File: aten/src/ATen/native/quantized/cpu/qconv_prepack.cpp [hunk 9]
```diff
@@ -127,7 +133,112 @@ class QConvPackWeightInt8 final : public c10::OperatorKernel {
...
+                              (uint8_t*)weight_contig.data_ptr<c10::quint8>(),
+                              (int32_t*)bias_contig.data_ptr<c10::qint32>()),
+                          {kernel_h, kernel_w},
+                          weight.q_scale(),
+                          weight.q_zero_point()});
+
+    return cpp_custom_type_hack::create(std::move(wt_ptr), weight.options());
+  }
+#endif // USE_PYTORCH_QNNPACK
+#if defined(USE_FBGEMM) || defined(USE_PYTORCH_QNNPACK)
```
Comment (dzhulgakov) [inline aten/src/ATen/native/quantized/cpu/qconv_prepack.cpp]: same comment

File: test/test_quantized.py [hunk 10]
```diff
@@ -1322,6 +1323,129 @@ def test_qlinear_qnnpack(self, batch_size, input_channels, output_channels, use_
...
+            pad_w,
+            dilation_h,
+            X_scale,
+            X_zp,
+            W_scale,
+            W_zp,
+            Y_scale,
+            Y_zp,
+            use_relu):
+        torch.backends.quantized.engine = torch.qnnpack
```
Comment (dzhulgakov) [inline test/test_quantized.py]: since it's a global flag, you probably want to have a context manager to unset it. Otherwise it'd mess up other tests

File: test/test_quantized.py [hunk 11]
```diff
@@ -1322,6 +1323,129 @@ def test_qlinear_qnnpack(self, batch_size, input_channels, output_channels, use_
...
+           pad_w=st.integers(0, 2),
+           dilation_h=st.integers(1, 1),
+           X_scale=st.floats(1.2, 1.6),
+           X_zp=st.integers(0, 4),
+           W_scale=st.floats(0.2, 1.6),
+           W_zp=st.integers(2, 5),
+           Y_scale=st.floats(4.2, 5.6),
+           Y_zp=st.integers(0, 4),
+           use_relu=st.booleans())
+    def test_qconv_qnnpack(
```
Comment (dzhulgakov) [inline test/test_quantized.py]: shouldn't this test be guarded by qnnpack being present? (we might want to have a function for it)

**Review by dzhulgakov — CHANGES_REQUESTED** [review]
looking good, some comments

File: test/test_quantized.py (hunk 11 above)
Comment (supriyar) [reply test/test_quantized.py]: There are some checks at the top of the class which skip if QNNPACK has not been built.

File: test/test_quantized.py (hunk 10 above)
Comment (supriyar) [reply test/test_quantized.py]: Do you have an example of how this can be done?

File: aten/src/ATen/native/quantized/cpu/qconv.cpp (hunk 8 above)
Comment (supriyar) [reply aten/src/ATen/native/quantized/cpu/qconv.cpp]: Yes, we can add that once we get the basic building blocks done :)

File: aten/src/ATen/native/quantized/cpu/qconv.cpp (hunk 7 above)
Comment (supriyar) [reply aten/src/ATen/native/quantized/cpu/qconv.cpp]: Sounds good. Removed the ifdef.

File: aten/src/ATen/native/quantized/cpu/qconv.cpp (hunk 6 above)
Comment (supriyar) [reply aten/src/ATen/native/quantized/cpu/qconv.cpp]: Yes, added a check in the prepack function.

File: test/test_quantized.py (hunk 10 above)
Comment (dzhulgakov) [reply test/test_quantized.py]: Something like this:
https://github.com/pytorch/pytorch/blob/21ba320cd55bc6eecc280e18c1d8d1212a9737f0/test/jit_utils.py#L509-L515

In general, if the tests would be valid for both fbgemm and qnnpack, we can also make the entire Test class to set the engine in the beginning and instantiate it twice for each engine (potentially with some skips if some tests aren't supported)

File: aten/src/ATen/native/quantized/cpu/qconv.cpp [hunk 12]
```diff
@@ -244,23 +259,147 @@ class QConv2dInt8 final : public c10::OperatorKernel {
...
+        std::all_of(
+            outShape.begin(), outShape.end(), [](int64_t i) { return i > 0; }),
+        "quanitzed::conv2d (qnnpack): each dimension of output tensor should be greater "
+        "than 0")
     TORCH_CHECK(
-        false,
-        "This PyTorch installation was not built "
-        "with FBGEMM operators");
+        (outShape[3] == out_ch),
+        "quanitzed::conv2d (qnnpack): Number of filters must be equal to number of "
```
Comment (dzhulgakov) [inline aten/src/ATen/native/quantized/cpu/qconv.cpp]: quantized typo in many places

Comment (supriyar) [reply aten/src/ATen/native/quantized/cpu/qconv.cpp]: I think I was typing in my sleep :)

File: test/test_quantized.py [hunk 13]
```diff
@@ -1323,6 +1322,129 @@ def test_qlinear_qnnpack(self, batch_size, input_channels, output_channels, use_
             np.testing.assert_equal(
                 Y_q_ref2.int_repr().numpy(), Y_q.int_repr().numpy())
 
+    @given(batch_size=st.integers(1, 3),
```
Comment (dzhulgakov) [inline test/test_quantized.py]: why do we need separate tests for qnnpack? can't we just use the same set of tests as for fbgemm? (and iterate over available engines in the beginning)

same applies to linear tests above

File: test/test_quantized.py [hunk 14]
```diff
@@ -1323,6 +1322,129 @@ def test_qlinear_qnnpack(self, batch_size, input_channels, output_channels, use_
...
+            pad_w,
+            dilation_h,
+            X_scale,
+            X_zp,
+            W_scale,
+            W_zp,
+            Y_scale,
+            Y_zp,
+            use_relu):
+        with enable_mobile_quantized_engine():
```
Comment (dzhulgakov) [inline test/test_quantized.py]: nit: I'd rather rename it to a function taking qengine as an argument, but since it's a test - it's minor

**Review by dzhulgakov — APPROVED** [review]
I guess one of us would need to rebase depending on whether permute PR lands first or not

File: aten/src/ATen/native/quantized/cpu/qconv_prepack.cpp [hunk 15]
```diff
@@ -125,19 +131,117 @@ class QConvPackWeightInt8 final : public c10::OperatorKernel {
...
+        out_ch,
+        weight.q_zero_point(),
+        weight.q_scale(),
+        std::numeric_limits<uint8_t>::min(),
+        std::numeric_limits<uint8_t>::max());
+
+    auto weight_contig = weight.contiguous();
+    auto bias_contig = bias.contiguous();
+    auto wt_ptr =
+        guts::make_unique<PackedConvWeightsQnnp>(PackedConvWeightsQnnp{
```
Comment (raghuramank100) [inline aten/src/ATen/native/quantized/cpu/qconv_prepack.cpp]: Where is the weight converted from int8 to uint8?

File: test/test_quantized.py (hunk 13 above)
Comment (raghuramank100) [reply test/test_quantized.py]: Ensuring numeric consistency of QNNPACK and FBGEMM is also a requirement for quantization aware training to work, so it will be good to check that both kernels pass the same tests. For QNNPACK we should also add tests that allow for full use of dynamic range of wts and activations

**Review by raghuramank100 — CHANGES_REQUESTED** [review]
Few comments/questions. Thanks!

File: aten/src/ATen/native/quantized/cpu/qconv_prepack.cpp (hunk 15 above)
Comment (supriyar) [reply aten/src/ATen/native/quantized/cpu/qconv_prepack.cpp]: That is in a separate PR -https://github.com/pytorch/pytorch/pull/26307

File: test/test_quantized.py (hunk 13 above)
Comment (supriyar) [reply test/test_quantized.py]: Discussed offline. Will merge tests in a different PR.
